Sprint 16

KudryashovAlexander opened the pull request:
Привет!
Вроде все сделано по заданию

KudryashovAlexander committed:
Добавлен WelcomePageViewController

KudryashovAlexander committed:
Добавлен вью и вьюМодель для доавбения категории

KudryashovAlexander committed:
Оптимизировал повторяющийся код, связанный  с UIButton

KudryashovAlexander committed:
Поправлено делегат добавления категорий

KudryashovAlexander committed:
поправлено вью с расписанием

KudryashovAlexander committed:
Добавлен метод удаления категорий

KudryashovAlexander committed:
Добавлено редактирования категории

KudryashovAlexander committed:
Добавлены  viewModel для Cell на вью конфигурации трекера

KudryashovAlexander committed:
Добавлено отслеживание количества символов в наименовании трекера

KudryashovAlexander committed:
поправлена кнопка добавления трекера

LizaIlinykh commented on Tracker/CategoryView/CategoriesViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+final class CategoriesViewModel {
+    
+    @Observable
+    private(set) var cateories = [CategoryViewModel]()
```
давай поправим опечатку
надо нажать правой кнопкой мыши -> refactor -> rename

KudryashovAlexander replied:
Ок

LizaIlinykh commented on Tracker/CategoryView/CategoriesViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+    func deleteCategory(index: Int) {
+        
+        let categoryviewModel = cateories[index]
+        try! trackerCatStory.deleteCategory(TrackerCategory(name: categoryviewModel.categoryName, trackers: []))
```
давай избавимся от принудительной операции и обработаем ошибку

KudryashovAlexander replied:
ок

LizaIlinykh commented on Tracker/CategoryView/CategoryAddViewController.swift:
```diff
@@ -0,0 +1,131 @@
+
+import UIKit
+
+class CategoryAddViewController: UIViewController {
```
давай обозначим как final

KudryashovAlexander replied:
ок

LizaIlinykh commented on Tracker/CategoryView/CategoryAddViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+        self.oldNCategoryName = oldNCategoryName
+    }
+    
+    private let trackerCategoryStore = TrackerCategoryStory()
```
свойства должны идти перед инициализаторами
вот ссылка на общепринятый формат оформления классов
https://habr.com/ru/articles/744470/

KudryashovAlexander replied:
пропустил, поправлю

LizaIlinykh commented on Tracker/CategoryView/CategoryAddViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+    func saveCategory() {
+        guard let categoryName = categoryName else { return }
+        let trackerCategory = TrackerCategory(name: categoryName, trackers: [])
+        try! trackerCategoryStore.addCategory(trackerCategory)
```
давай избавимся от принудительной операции и обработаем ошибку

KudryashovAlexander replied:
ок

LizaIlinykh commented on Tracker/CategoryView/CategoryAddViewModel.swift:
```diff
@@ -0,0 +1,50 @@
+    func renameCategory(newName: String) {
+        guard let oldName = oldNCategoryName else {return}
+        if newName != oldName {
+            try! trackerCategoryStore.changeNameCategory(oldName: oldName, newName: newName)
```
давай избавимся от принудительной операции и обработаем ошибку

KudryashovAlexander replied:
ок

LizaIlinykh commented on Tracker/CategoryView/CategoryTableViewCell.swift:
```diff
@@ -0,0 +1,81 @@
+
+import UIKit
+
+class CategoryTableViewCell: UITableViewCell {
```
final

KudryashovAlexander replied:
ок

LizaIlinykh commented on Tracker/CategoryView/CategoryViewController.swift:
```diff
@@ -0,0 +1,239 @@
+
+class CategoryViewController: UIViewController {
+    
+    var viewModel: CategoriesViewModel!
```
лучше занести через инициализатор закрыв протоколом
сделать ее let и private
https://swiftblog.org/vnedrenie-zavisimostej/

KudryashovAlexander replied:
вынес через иницилизатор +let +private

LizaIlinykh commented on Tracker/CategoryView/CategoryViewController.swift:
```diff
@@ -0,0 +1,239 @@
+        createCategoryButton.widthAnchor.constraint(equalToConstant: contentView.frame.width - 40).isActive = true
+        createCategoryButton.centerXAnchor.constraint(equalTo: contentView.centerXAnchor).isActive = true
+        createCategoryButton.heightAnchor.constraint(equalToConstant: 60).isActive = true
+    }
```
лучше активировать при помощи NSLayout.activate(массив констрейнтов которые нужно активировать)

KudryashovAlexander replied:
Ну ок

LizaIlinykh commented on Tracker/CategoryView/CategoryViewController.swift:
```diff
@@ -0,0 +1,239 @@
+                                      viewController: self) {
+            
+            self.viewModel.deleteCategory(index: index)
+        }
```
лучше отдать создание моделей вью модели, вьюха не должна создавать модельки

KudryashovAlexander replied:
Мне показалось логичным, чтобы Alert показывал ViewController. А сам метод удаления уже по вью модели

LizaIlinykh replied:
алерт может показывать контроллер, но модели для него должна создавать вью модель

LizaIlinykh commented on Tracker/TrackerConfigurationView/ConfigTableViewCell.swift:
```diff
@@ -0,0 +1,100 @@
+            guard let self = self,
+                  let name = name else { return }
+            self.selectedPropertyLabel.text = name
+            self.layoutConfiguration(selectedIsActive:true)
```
ты заново создаешь констрейнты каждый раз при изменении selectedPoperty
констрейнты стоит доавблять единожды, а затем менять их если это необходимо

KudryashovAlexander replied:
 поправил

KudryashovAlexander committed:
поправлено по критическим и по некритическим комментариям ревью.

Касаемо кнопки плюс на трекере возможно ранее был немного другой макет, могу ошибаться. Но не придумывал же я его сам)

KudryashovAlexander committed:
Поправлено по пул реквесту

LizaIlinykh commented on Tracker/CategoryView/CategoryViewController.swift:
```diff
@@ -9,8 +9,8 @@ import UIKit
-    var viewModel: CategoriesViewModel!
-    private var alertPresenter = AlertPresener()
+    private let viewModel: CategoriesViewModel
+    private let alertPresenter = AlertPresener()
```
будет круто все зависимости заносить извне, закрыв их протоколом. Потом будет удобно заменять эти обьекты при надобности
AlertPresenter - опечатка

KudryashovAlexander committed:
Поправлено отслеживание кнопки нажатия создания трекера

KudryashovAlexander committed:
правка

KudryashovAlexander committed:
Поправлены скроллинг и заглушка при поиске. То по критическим комментариям все поправил. Добавил ВьюМодель для трекерКонфигурейшн. Возможно не вся логика еще перенесена.
По миниправкам:
1) Расположение лейблов Emoji и Цвет поправил
2) Галочка у меня показывается. При переходе на экран категорий галочка вставляется. Если еще раз открыть категории то ранее выбранная категория показывается
3) согласно макету в коллекциях должно быть по шесть цветов / эмоджи в ряду - по вот этому комментарию не совсем понял. у меня как в макете - 6 в ряду.
На всякий случай записал как работает у меня: https://disk.yandex.lt/i/ODfezDHTauMPlQ

KudryashovAlexander committed:
поправлено после ревью
